(MODULES-3037) Manage sources

ferventcoder opened the pull request:
Provide the ability manage a source with a custom type and provider.
This allows a source to be ensurable, and also allows for a source to
be manageable in the same converge that Chocolatey is installed
during. This further allows for installing packages from custom
sources all in the same Puppet run.

ferventcoder commented on spec/unit/puppet_x/chocolatey/chocolatey_common_spec.rb:
```diff
@@ -0,0 +1,9 @@
+describe 'Chocolatey Common' do
+
+  context 'on Windows', :if => Puppet::Util::Platform.windows? do
+
```
This is empty right now, but I wanted to get eyes on this so we can get it merged and through soon. So we can pick this up in other tickets or as maint.

ferventcoder commented:
@puppetlabs/windows please review.

glennsarti commented on lib/puppet_x/chocolatey/chocolatey_common.rb:
```diff
@@ -0,0 +1,81 @@
+        if Puppet::Util::Platform.windows?
+          chocoInstallPath = PuppetX::Chocolatey::ChocolateyInstall.install_path
+          #default_location = $::choco_installpath || ENV['ALLUSERSPROFILE'] + '\chocolatey'
+          chocopath = ('C:\ProgramData\chocolatey' if file_exists?('C:\ProgramData\chocolatey\bin\choco.exe')) ||
```
I'm not 100% comfortable with using a hardcoded path for C:\ProgramData.  It tends to be surfaced as `%ALLUSERSPROFILE%` or at least `%SYSTEMDRIVE%\ProgramData`

ferventcoder replied:
This section is a straight copy from the package provider. There be voodo and reasons there, they haven't been all determined yet.

glennsarti commented on lib/puppet_x/chocolatey/chocolatey_common.rb:
```diff
@@ -0,0 +1,81 @@
+          #default_location = $::choco_installpath || ENV['ALLUSERSPROFILE'] + '\chocolatey'
+          chocopath = ('C:\ProgramData\chocolatey' if file_exists?('C:\ProgramData\chocolatey\bin\choco.exe')) ||
+              (chocoInstallPath if chocoInstallPath && file_exists?("#{chocoInstallPath}\\bin\\choco.exe")) ||
+              (ENV['ChocolateyInstall'] if ENV['ChocolateyInstall'] && file_exists?("#{ENV['ChocolateyInstall']}\\bin\\choco.exe")) ||
```
Wouldn't the ChocolateyInstall ENV var have the hightest priority and be at the beginning of this OR code?

ferventcoder replied:
There is a reason behind the order. I'm just not sure what it is.

ferventcoder replied:
It does have to do with suitability and first run.

glennsarti commented on lib/puppet_x/chocolatey/chocolatey_common.rb:
```diff
@@ -0,0 +1,81 @@
+      end
+      module_function :chocolatey_command
+
+      def set_env_chocolateyinstall
```
Seems to be some overlap between this;
https://github.com/puppetlabs/puppetlabs-chocolatey/blob/master/lib/puppet/provider/package/chocolatey.rb#L92-L94
lib/puppet_x/chocolatey/chocolatey_common.rb#L33-36
Unless I'm missing something?

ferventcoder replied:
No, not at all. The plan is to convert the package code over to using this. It was getting a little out of scope for this ticket as it was.

ferventcoder replied:
No more overlap.

glennsarti commented on spec/unit/puppet_x/chocolatey/chocolatey_common_spec.rb:
```diff
@@ -0,0 +1,63 @@
+    end
+
+    it "should be true when version is greater than 0.9.9.0" do
+      PuppetX::Chocolatey::ChocolateyCommon.expects(:choco_version).returns('0.9.9.0')
```
This appears to be a duplicate of the test above.  Is `.returns('0.9.9.0')` correct here?

ferventcoder replied:
should have been a higher number.

glennsarti commented:
Is it worth refactoring out the "magic" version numbers in the unit tests  e.g. 0.9.9.10 etc?

ferventcoder commented:
> Is it worth refactoring out the "magic" version numbers in the unit tests e.g. 0.9.9.10 etc?

Yeah, I think we should make it clearer - I'll change it to minimum_supported_version - that should make it pretty clear what is being tested.

glennsarti commented on spec/unit/puppet_x/chocolatey/chocolatey_install_spec.rb:
```diff
@@ -0,0 +1,28 @@
+
+  context 'on Linux', :if => Puppet.features.posix? do
+    it "should return the default install path on a non-windows system" do
+      PuppetX::Chocolatey::ChocolateyInstall.install_path.must == 'C:\ProgramData\chocolatey'
```
This seems counter-intuitive.  On a linux system wouldn't the install path be nil or empty?

ferventcoder replied:
Default path when one is not found.

glennsarti commented on spec/unit/puppet_x/chocolatey/chocolatey_common_spec.rb:
```diff
@@ -0,0 +1,67 @@
+
+  context ".choco_version" do
+    it "should return PuppetX::Chocolatey::ChocolateyVersion.version" do
+      expected = '0.9.9.0.1'
```
Are you using a 5 segment version here on purpose?  Seems most of the version for chocolatey are 4 segment ala SemVer

ferventcoder replied:
Yes, it's so that the verification is valid. I could accidentally have a false positive on a version if the version was a real one.

glennsarti commented on spec/unit/puppet/type/chocolateysource_spec.rb:
```diff
@@ -0,0 +1,138 @@
+      it "should not accept other string values" do
+        expect {
+          resource[param_symbol] = 'value'
+        }.to raise_error(Puppet::Error, /An integer is necessary for priority/)
```
I'm don't think you should have the word `priority` in the regex.  What if you added another numeric parameter?

ferventcoder replied:
Good point. That was a mistake.

glennsarti commented on spec/unit/puppet/type/chocolateysource_spec.rb:
```diff
@@ -0,0 +1,138 @@
+      it "should not accept symbol values" do
+        expect {
+          resource[param_symbol] = :whenever
+        }.to raise_error(Puppet::Error, /An integer is necessary for priority/)
```
I don't think you should have the word `priority` in the regex.  What if you added another numeric parameter?

ferventcoder replied:
Thanks for noticing that.

glennsarti commented on spec/unit/puppet/type/chocolateysource_spec.rb:
```diff
@@ -0,0 +1,138 @@
+      end
+
+      it "should accept any string that represents a numeric value" do
+        resource[param_symbol] = '1'
```
Do you need to test for negative integers?  e.g. -1

ferventcoder replied:
It will allow it.

ferventcoder commented:
@puppetlabs/windows paired with @glennsarti some today on this as we fought through refactoring the Chocolatey package provider bits to get them to use `PuppetX::Chocolatey::ChocolateyCommon`. Lots of tests needed touched and adjusted. 

There may be an underlying issue or two with when choco.exe is actually found (every system that runs these tests has Chocolatey installed). We removed it for a bit to ensure things would work without it.

ferventcoder commented:
I'll pick these up tomorrow and work to get them overridden.

ferventcoder commented:
@puppetlabs/windows all comments have been addressed. Ready to go.

glennsarti commented:
Looks good to me.
